### DISCUSSION TITLE
"Signature of a method incompatible with supertype" vs pytest fixtures

### DISCUSSION THREAD

**Discussion by serhiy-storchaka** [pytest-dev/pytest#10074] [General]
I don't know whether it is idiomatic or not, but I use the following approach to parametrize `pytest` tests. Tests are defined as method in a class. Parameters are defined as fixtures in a class. Fixtures are overridden in subclasses to run tests with different parameters.

For example:
```python
import pytest


class TestA:
    @pytest.fixture
    def foo(self) -> int:
        return 42

    def test_foo(self, foo: int) -> None:
        assert foo > 0


@pytest.fixture
def bar() -> str:
    return "abc"


class TestB(TestA):
    @pytest.fixture
    def foo(self, bar: str) -> int:  # line 20
        return len(bar)
```
`TestA.test_foo` and `TestB.test_foo` are executed with different values of `foo`. The problem is that fixtures can depend on other fixtures, and signatures for methods defining fixture can be different. MyPy complains:
```
t.py:20: error: Signature of "foo" incompatible with supertype "TestA"
t.py:20: note:      Superclass:
t.py:20: note:          def foo(self) -> int
t.py:20: note:      Subclass:
t.py:20: note:          def foo(self, bar: str) -> int
```
It is not real error in the user code that the methods have different signatures. These methods are not called directly. They are called by pytest which provides all arguments.

I can add "# type: ignore", but I do not want to lose static type checking completely. It is useful to check that the type of the parameter matches the returning type of corresponding fixture function.

mypy 0.961 (compiled: yes)
pytest 7.1.2
Python 3.9.13

**Comment by The-Compiler** [comment]
This doesn't seem like something pytest can do anything about. Maybe it should be a discussion rather than an issue?

I've seen this message being disputed for other reasons too:

- https://github.com/python/mypy/issues/1237

Note that mypy [supports](https://mypy.readthedocs.io/en/stable/common_issues.html#spurious-errors-and-locally-silencing-the-checker) adding [error codes](https://mypy.readthedocs.io/en/stable/error_codes.html#error-codes) to type ignores.

**Reply by serhiy-storchaka** [reply]
At least partially it can be solved from pytest side. Changing the return type of `pytest.fixture` from `Union[FixtureFunctionMarker, FixtureFunction]` to `Union[FixtureFunctionMarker, Callable[..., Any]]` silences MyPy. Of course it is at the cost of losing some information, but is this information used anywhere?

**Reply by RonnyPfannschmidt** [reply]
@nicoddemus @The-Compiler i wonder if we should finally remove FixtureFunction for a FixtureDefinition (which is not callable)

the deprecation errors since a while now and its about timme we take it away - that way fixture definitions will no longer be mistaken as method definitions

**Reply by nicoddemus** [reply]
Deprecation errors or warnings you mean?

**Reply by RonnyPfannschmidt** [reply]
@nicoddemus we hard fail with a pytest.fail these days

**Reply by nicoddemus** [reply]
Ahh then sure, I'm 👍
